File: server/rhs.go
// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package main

import (
	"strings"
	"time"

	jira "github.com/andygrunwald/go-jira"
	"github.com/pkg/errors"

	"github.com/mattermost/mattermost-plugin-jira/server/utils/kvstore"
	"github.com/mattermost/mattermost-plugin-jira/server/utils/types"
)

const rhsIssuesPageSize = 20

var rhsSearchFields = []string{
	"summary", "status", "issuetype", "updated", "project",
}

// rhsCloudClient is the Cloud-only RHS API. Do not add these methods to Client.
type rhsCloudClient interface {
	rhsStatusLister
	SearchJQL(params CloudSearchParams) (*CloudSearchResult, error)
}

func (p *Plugin) loadRHSInstance(instanceID types.ID) (Instance, error) {
	if instanceID == "" {
		return nil, errors.Wrap(ErrInvalidRHSTab, "instance_id is required")
	}
	instance, err := p.instanceStore.LoadInstance(instanceID)
	if err != nil {
		if errors.Is(err, kvstore.ErrNotFound) {
			return nil, errors.Wrap(ErrInvalidRHSTab, "unknown instance_id")
		}
		return nil, err
	}
	return instance, nil
}

func (p *Plugin) resolveRHSUserClient(instanceID, mattermostUserID types.ID) (Instance, rhsCloudClient, error) {
	instance, err := p.loadRHSInstance(instanceID)
	if err != nil {
		return nil, nil, err
	}

	client, instance, _, err := p.getClient(instance.GetID(), mattermostUserID)
	if err != nil {
		return nil, nil, err
	}
	rhs, ok := client.(rhsCloudClient)
	if !ok {
		return nil, nil, ErrRHSNotCloud
	}
	return instance, rhs, nil
}

func (p *Plugin) resolveRHSAdminClient(instance Instance, adminUserID types.ID) (rhsCloudClient, error) {
	switch inst := instance.(type) {
	case *cloudInstance:
		raw, err := inst.getClientForBot()
		if err != nil {
			return nil, err
		}
		rhs, ok := newCloudClient(raw).(rhsCloudClient)
		if !ok {
			return nil, ErrRHSNotCloud
		}
		return rhs, nil

	case *cloudOAuthInstance:
		connection, err := p.userStore.LoadConnection(inst.GetID(), adminUserID)
		if err != nil {
			return nil, err
		}
		client, err := inst.GetClient(connection)
		if err != nil {
			return nil, err
		}
		rhs, ok := client.(rhsCloudClient)
		if !ok {
			return nil, ErrRHSNotCloud
		}
		return rhs, nil

	default:
		return nil, ErrRHSNotCloud
	}
}

func rhsBrowseURL(instance Instance, issueKey string) string {
	return strings.TrimRight(instance.GetJiraBaseURL(), "/") + "/browse/" + issueKey
}

func rhsTimeRFC3339(t jira.Time) string {
	tt := time.Time(t)
	if tt.IsZero() {
		return ""
	}
	return tt.UTC().Format(time.RFC3339)
}

func normalizeRHSIssue(instance Instance, issue jira.Issue) RHSIssue {
	out := RHSIssue{
		Key:       issue.Key,
		BrowseURL: rhsBrowseURL(instance, issue.Key),
	}
	if issue.Fields == nil {
		return out
	}
	f := issue.Fields
	out.Summary = f.Summary
	out.IssueType = f.Type.Name
	out.IssueTypeIconURL = f.Type.IconURL
	out.Project = f.Project.Key
	out.Updated = rhsTimeRFC3339(f.Updated)
	if f.Status != nil {
		out.Status = RHSIssueStatus{
			Name:        f.Status.Name,
			CategoryKey: f.Status.StatusCategory.Key,
		}
	}
	return out
}

func normalizeRHSIssues(instance Instance, issues []jira.Issue) []RHSIssue {
	out := make([]RHSIssue, 0, len(issues))
	for _, issue := range issues {
		out = append(out, normalizeRHSIssue(instance, issue))
	}
	return out
}

func pickRHSTab(tabs []RHSTabEntry, identity string) RHSTabEntry {
	parsed, err := parseRHSTabIdentity(identity)
	if err != nil {
		return rhsAssignedTab()
	}
	want := rhsTabIdentity(parsed)
	for _, tab := range tabs {
		if rhsTabIdentity(tab) == want {
			return tab
		}
	}
	return rhsAssignedTab()
}

type rhsIssuesResult struct {
	Issues        []RHSIssue    `json:"issues"`
	Tabs          []RHSTabEntry `json:"tabs"`
	NextPageToken string        `json:"nextPageToken"`
	IsLast        bool          `json:"isLast"`
}

func (p *Plugin) getRHSIssues(instanceID, mattermostUserID types.ID, tabIdentity, sort, nextPageToken string) (*rhsIssuesResult, error) {
	instance, client, err := p.resolveRHSUserClient(instanceID, mattermostUserID)
	if err != nil {
		return nil, err
	}

	entry, err := p.getInstanceStatuses(instance.GetID(), mattermostUserID, client)
	if err != nil {
		return nil, err
	}

	conf := p.getConfig()
	configured := conf.RHSStatusTabs[string(instance.GetID())]
	tabs := resolveTabs(configured, entry.statuses, entry.categories)

	selected := pickRHSTab(tabs, tabIdentity)
	if sort == "" {
		sort = rhsDefaultSort
	}

	jql, err := buildTabJQL(selected, sort, validCategoryKeysFrom(entry.categories))
	if err != nil {
		return nil, err
	}

	search, err := client.SearchJQL(CloudSearchParams{
		JQL:           jql,
		Fields:        rhsSearchFields,
		MaxResults:    rhsIssuesPageSize,
		NextPageToken: nextPageToken,
	})
	if err != nil {
		return nil, err
	}

	return &rhsIssuesResult{
		Issues:        normalizeRHSIssues(instance, search.Issues),
		Tabs:          tabs,
		NextPageToken: search.NextPageToken,
		IsLast:        search.IsLast,
	}, nil
}

type rhsStatusesResult struct {
	Statuses   []*JiraStatus         `json:"statuses"`
	Categories []*JiraStatusCategory `json:"categories"`
}

func (p *Plugin) getRHSStatuses(instanceID, adminUserID types.ID) (*rhsStatusesResult, error) {
	instance, err := p.loadRHSInstance(instanceID)
	if err != nil {
		return nil, err
	}

	client, err := p.resolveRHSAdminClient(instance, adminUserID)
	if err != nil {
		return nil, err
	}

	entry, err := p.getInstanceStatuses(instance.GetID(), rhsAdminStatusCacheUserID(instance, adminUserID), client)
	if err != nil {
		return nil, err
	}
	p.enrichStatusesWithProjects(client, entry.statuses)
	return &rhsStatusesResult{
		Statuses:   entry.statuses,
		Categories: entry.categories,
	}, nil
}

File: webapp/src/types/rhs.ts
// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export type RHSErrorCode =
    'not_connected' |
    'rate_limited' |
    'not_authorized' |
    'not_cloud' |
    'invalid_request' |
    'internal_error';

export type RHSSort = 'updated' | 'created';

export type RHSTab =
    | {kind: 'assigned'; name: string}
    | {kind: 'category'; key: string; name: string}
    | {kind: 'status'; id: string; name: string};

export type RHSIssueStatus = {
    name: string;
    categoryKey: string;
};

export type RHSIssue = {
    key: string;
    summary: string;
    browseUrl: string;
    status: RHSIssueStatus;
    issueType: string;
    issueTypeIconUrl?: string;
    project: string;
    updated: string;
};

export type RHSIssuesResponse = {
    issues: RHSIssue[];
    tabs: RHSTab[];
    nextPageToken: string;
    isLast: boolean;
};

export type RHSStatusCategory = {
    id: number;
    key: string;
    name: string;
};

export type RHSStatusProject = {
    id?: string;
    key?: string;
    name?: string;
};

export type RHSStatus = {
    id: string;
    name: string;
    statusCategory: RHSStatusCategory;
    project?: RHSStatusProject;
};

export type RHSStatusesResponse = {
    statuses: RHSStatus[];
    categories: RHSStatusCategory[];
};

export type RHSViewState = {
    instance: string;
    tab: RHSTab;
    sort: RHSSort;
};

export const RHS_DEFAULT_SORT: RHSSort = 'updated';

export const RHS_ASSIGNED_TAB_NAME = 'Assigned to me';

export const RHS_DEFAULT_TAB: RHSTab = {
    kind: 'assigned',
    name: RHS_ASSIGNED_TAB_NAME,
};

export type FetchRHSIssuesArgs = {
    instanceID: string;
    tab: RHSTab;
    sort: RHSSort;
};

export type GetRHSIssuesParams = FetchRHSIssuesArgs & {
    nextPageToken?: string;
};

export function rhsTabIdentity(tab: RHSTab): string {
    switch (tab.kind) {
    case 'assigned':
        return 'assigned';
    case 'category':
        return 'category:' + tab.key;
    case 'status':
        return 'status:' + tab.id;
    default: {
        const exhaustive: never = tab;
        return exhaustive;
    }
    }
}

export function rhsTabFromAPI(raw: {kind?: unknown; name?: unknown; key?: unknown; id?: unknown}): RHSTab | null {
    const name = typeof raw.name === 'string' && raw.name ? raw.name : '';
    switch (raw.kind) {
    case 'assigned':
        return {
            kind: 'assigned',
            name: RHS_ASSIGNED_TAB_NAME,
        };
    case 'category': {
        if (typeof raw.key !== 'string' || !raw.key) {
            return null;
        }
        return {
            kind: 'category',
            key: raw.key,
            name: name || raw.key,
        };
    }
    case 'status': {
        if (typeof raw.id !== 'string' || !raw.id) {
            return null;
        }
        return {
            kind: 'status',
            id: raw.id,
            name: name || raw.id,
        };
    }
    default:
        return null;
    }
}
